File: 1029 Two City Scheduling.py
class Solution:
    def twoCitySchedCost(self, costs: List[List[int]]) -> int:
        costs.sort(key = lambda x: x[0] - x[1])
        return sum(a[0] for a in costs[:len(costs)//2]) + sum(b[1] for b in costs[len(costs)//2:])

'''
# One-line solution
class Solution:
    def twoCitySchedCost(self, costs: List[List[int]]) -> int:
        return sum([v[i // (len(costs)//2)] for i, v in enumerate(sorted(costs, key=lambda x: x[0] - x[1]))])

class Solution:
    def twoCitySchedCost(self, costs: List[List[int]]) -> int:
        alt = sorted([(a-b, a, b) for a,b in costs])
        return sum(it[1] for it in alt[:len(costs)//2]) + sum(it[2] for it in alt[len(costs)//2:])  
'''

File: 1029 Two City Scheduling.java
class Solution {
    public int twoCitySchedCost(int[][] costs) {
        int len = costs.length;
        Comparator cmp = new CityComparator();
        Arrays.sort(costs, cmp);
        int sum = 0;
        for (int i = 0; i < len; i++) {
            sum += (i < len / 2)? costs[i][0] :costs[i][1];
        }
        return sum;
    }
}
class CityComparator implements Comparator<int[]> {
    @Override
    public int compare(int[] o1, int[] o2) {
        return (o1[0] - o1[1]) - (o2[0] - o2[1]);
    }
}

/* Faster that using diff
class Solution {
    public int twoCitySchedCost(int[][] costs) {
        int[] diff = new int[costs.length];
        int sum = 0;
        for(int i=0; i<costs.length; i++) {
            diff[i] = costs[i][0] - costs[i][1];
            sum += costs[i][0];
        }
        Arrays.sort(diff);
        for(int i=diff.length/2; i<diff.length; i++) {
            sum -= diff[i];
        }
        return sum;
    }
}
*/